bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/cipher/SymmetricKeyCipher.java

```diff
@@ -0,0 +1,167 @@
+
+    /**
+     * Activate will try to load the encryption key. If an existing encryption key does not exists,
+     * it will generate a new one and save to {@code org.osgi.service.cm.ConfigurationAdmin}
+     *
+     * @throws NoSuchAlgorithmException When encryption algorithm is not available {@code #ENCRYPTION_ALGO}
+     * @throws IOException if access to persistent storage fails (@code org.osgi.service.cm.ConfigurationAdmin)
+     */
+    @Activate
+    public void activate() throws NoSuchAlgorithmException, IOException {
+        // load or generate the encryption key
+        encryptionKey = getOrGenerateEncryptionKey(configuration);
+    }
+
+    @Override
+    public String getUniqueCipherId() {
+        return CIPHER_ID;
+    }
+
+    @Override
+    public String encrypt(@Nullable String plainText) throws GeneralSecurityException {
+        if (plainText == null) {
+            return null;
+        }
+
+        // Generate IV
+        byte iv[] = new byte[IV_BYTE_SIZE];
+        random.nextBytes(iv);
+        cipherEnc = Cipher.getInstance(ENCRYPTION_ALGO);
+        cipherEnc.init(Cipher.ENCRYPT_MODE, encryptionKey, new IvParameterSpec(iv));
+        byte[] encryptedBytes = cipherEnc.doFinal(plainText.getBytes(StandardCharsets.UTF_8));
+        byte[] encryptedBytesWithIV = new byte[encryptedBytes.length + IV_BYTE_SIZE];
+
+        // copy iv to the start of array
+        System.arraycopy(iv, 0, encryptedBytesWithIV, 0, IV_BYTE_SIZE);
+        // append encrypted text to tail
+        System.arraycopy(encryptedBytes, 0, encryptedBytesWithIV, IV_BYTE_SIZE, encryptedBytes.length);
+        String encryptedBase64String = Base64.getEncoder().encodeToString(encryptedBytesWithIV);
+
+        return encryptedBase64String;
+    }
+
+    @Override
+    public String decrypt(@Nullable String base64CipherText) throws GeneralSecurityException {
+        if (base64CipherText == null) {
+            return null;
+        }
+        // base64 decode the base64CipherText
+        byte[] decodedCipherTextWithIV = Base64.getDecoder().decode(base64CipherText);
+        // Read IV
+        byte[] iv = new byte[IV_BYTE_SIZE];
+        System.arraycopy(decodedCipherTextWithIV, 0, iv, 0, IV_BYTE_SIZE);
+
+        byte[] cipherTextBytes = new byte[decodedCipherTextWithIV.length - IV_BYTE_SIZE];
+        System.arraycopy(decodedCipherTextWithIV, IV_BYTE_SIZE, cipherTextBytes, 0, cipherTextBytes.length);
+
+        Cipher cipherDec = Cipher.getInstance(ENCRYPTION_ALGO);
+        cipherDec.init(Cipher.DECRYPT_MODE, encryptionKey, new IvParameterSpec(iv));
+        byte[] decryptedBytes = cipherDec.doFinal(cipherTextBytes);
+        return new String(decryptedBytes, StandardCharsets.UTF_8);
+    }
+
+    private static SecretKey generateEncryptionKey() throws NoSuchAlgorithmException {
+        KeyGenerator keygen = KeyGenerator.getInstance("AES");
+        keygen.init(128); // do not use very high grade encryption due to export limit
```

Comment:
Could you replace the magic number 128 by a constant?

Reply:
agreed